### PR TITLE
QNN Google Bert & OpenAI Clip Olive Recipe

### PR DESCRIPTION
**Pull request by gtonpe**
## Describe your changes
QNN Google Bert & OpenAI Clip Olive Recipe (fp32 and W8A16)

## Checklist before requesting a review
- [ ] Add unit tests for this change.
- [ ] Make sure all tests can pass.
- [ ] Update documents if necessary.
- [ ] Lint and apply fixes to your code by running `lintrunner -a`
- [ ] Is this a user-facing change? If yes, give a description of this change to be included in the release notes.
- [ ] Is this PR including examples changes? If yes, please remember to update [example documentation](https://github.com/microsoft/Olive/blob/main/docs/source/examples.md) in a follow-up PR.

### CODE REVIEW

**Comment by xiaoyu-work** [comment]
Can you follow this instruction to format the code? https://github.com/microsoft/Olive/blob/main/CONTRIBUTING.md

**Comment by xiaoyu-work** [comment]
Can you also update README.md with the examples you added?

File: examples/bert/google_bert_qnn_fp32.json [hunk 1]
```diff
@@ -0,0 +1,177 @@
+{
+    "input_model": {
+        "type": "PyTorchModel",
+        "model_path": "google-bert/bert-base-multilingual-cased",
+        "io_config": {
+            "input_names": [ "input_ids", "attention_mask", "token_type_ids" ],
+            "input_shapes": [ [ 1, 128 ], [ 1, 1, 128, 128 ], [ 1, 128 ] ],
+            "input_types": [ "int32", "float32", "int32" ],
+            "output_names": [ "logits" ],
+            "dynamic_axes": {
```
Comment (jambayk) [inline examples/bert/google_bert_qnn_fp32.json:10]: if dynamic axis from io_config is removed, the exported model will have static shape and we won't need the dynamic to static shape pass

File: examples/bert/google_bert_qnn_fp32.json [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                "input_ids": { "0": "batch_size", "1": "sequence_length" },
+                "attention_mask": { "0": "batch_size", "2": "sequence_length", "3": "sequence_length" },
+                "token_type_ids": { "0": "batch_size", "1": "sequence_length" },
+                "logits": { "0": "batch_size" }
+            }
+        },
+        "model_loader": "load_bert_nsp_model",
```
Comment (jambayk) [inline examples/bert/google_bert_qnn_fp32.json:17]: is this custom model loader only so that the attention mask is 4d? from my earlier investigations, I experienced accuracy drop with 2d mask because the mask value was too large to quantize correctly. this is the reason why we added the `ReplaceAttentionMaskValue` graph surgery. do you experience accuracy issues with a 2d mask even after the surgery?

File: examples/bert/google_bert_qnn_fp32.json [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+                "type": "tokenize_dataset",
+                "input_cols": [ "sentence1", "sentence2" ],
+                "label_col": "label",
+                "seq_length": 128,
+                "model_name": "google-bert/bert-base-multilingual-cased"
+            },
+            "dataloader_config": { "batch_size": 1 }
+        },
+        {
+            "name": "wiki_data",
```
Comment (jambayk) [inline examples/bert/google_bert_qnn_fp32.json:48]: i don't see the `wiki_data` and `eval_data` being used in this workflow elsewhere

File: examples/bert/google_bert_qnn_fp32.json (hunk 2 above)
Comment (jambayk) [reply examples/bert/google_bert_qnn_fp32.json:17]: the large negative value is replace with `-1e4` by default but you can configure it if you want a different value like `-50`

`{ "surgeon": "ReplaceAttentionMaskValue", "replacement" -50 }`

Comment (jambayk) [reply examples/bert/google_bert_qnn_fp32.json:17]: Oh i see, the model is a nextsequenceprediction model for which we don't have a corresponding task. but I am still curious if a 4d mask is needed and if the attention mask surgery is enough. otherwise, we would require the model user to have to create the 4d mask themselves when running the model.

File: examples/bert/bert_script.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
+from collections import OrderedDict
```
[the remark this answers is not in the document]
Comment (gtonpe) [reply examples/bert/bert_script.py:1]: done

File: examples/bert/google_bert_qnn_fp32.json (hunk 1 above)
Comment (gtonpe) [reply examples/bert/google_bert_qnn_fp32.json:10]: done, thanks for the suggestion

File: examples/bert/google_bert_qnn_fp32.json (hunk 2 above)
Comment (gtonpe) [reply examples/bert/google_bert_qnn_fp32.json:11]: We are going with 4d mask since accuracy issues are observed with 2d mask

File: examples/bert/google_bert_qnn_fp32.json (hunk 3 above)
Comment (gtonpe) [reply examples/bert/google_bert_qnn_fp32.json:27]: We have addresses this in the latest commit

File: examples/bert/google_bert_qnn.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
+import sys
+from collections import OrderedDict
+from pathlib import Path
+from typing import List, Optional
+
+import torch
+
+sys.path.append(Path.cwd().as_posix().replace("/", "\\"))
```
Comment (jambayk) [inline examples/bert/google_bert_qnn.py:8]: `data_config` also has a script_dir option so this should not be needed,

Comment (gtonpe) [reply examples/bert/google_bert_qnn.py:8]: in our environment, this line was needed to find other modules, otherwise we were getting "no module found" error

File: examples/clip/openai_clip-vit-base-patch16_ptq_qnn_qdq.json [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+        }
+    },
+    "passes": {
+        "conversion": { "type": "OnnxConversion", "target_opset": 20 },
+        "surgery": { "type": "GraphSurgeries", "surgeries": [ { "surgeon": "ReplaceAttentionMaskValue" } ] },
+        "quantization": {
+            "type": "OnnxStaticQuantization",
+            "quant_preprocess": true,
+            "data_config": "quant_data_config",
+            "op_types_to_quantize": [
```
Comment (jambayk) [inline examples/clip/openai_clip-vit-base-patch16_ptq_qnn_qdq.json:180]: We have an `op_types_to_exclude` option https://github.com/microsoft/Olive/blob/69e3de54667701dc7d21f149cabd86d17a3fffb9/olive/passes/onnx/quantization.py#L57 which I think is a better option to use here since looks like we are including almost all ops?

File: examples/bert/google_bert_qnn.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+from bert_common import (
```
[the remark this answers is not in the document]
Comment (gtonpe) [reply examples/bert/google_bert_qnn.py:6]: Fixed

File: examples/bert/google_bert_qnn.py [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+    SimpleBert,
+    npz_to_hfdataset,
+    tokenize_hfdataset2,
+)
+from datasets import load_dataset
```
[the remark this answers is not in the document]
Comment (gtonpe) [reply examples/bert/google_bert_qnn.py:11]: fixed

File: examples/bert/google_bert_qnn.py [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+from transformers import (
+    AutoModelForNextSentencePrediction as AutoModelNSP,
+)
+from transformers import (
```
[the remark this answers is not in the document]
Comment (gtonpe) [reply examples/bert/google_bert_qnn.py:15]: fixed

File: examples/bert/google_bert_qnn.py [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+    AutoModelForSequenceClassification as AutoModelSCL,
+)
+from transformers import (
```
[the remark this answers is not in the document]
Comment (gtonpe) [reply examples/bert/google_bert_qnn.py:18]: fixed

File: examples/bert/google_bert_qnn.py [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    AutoTokenizer,
+    BertModel,
+)
+from wikitext import create_nsp_dataset
```
[the remark this answers is not in the document]
Comment (gtonpe) [reply examples/bert/google_bert_qnn.py:22]: fixed

File: examples/bert/google_bert_qnn.py [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+
+from olive.data.component.dataset import BaseDataset
```
[the remark this answers is not in the document]
Comment (gtonpe) [reply examples/bert/google_bert_qnn.py:24]: fixed

File: examples/bert/google_bert_qnn.py [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+from olive.data.registry import Registry
```
[the remark this answers is not in the document]
Comment (gtonpe) [reply examples/bert/google_bert_qnn.py:25]: fixed

File: examples/bert/google_bert_qnn.py (hunk 5 above)
Comment (gtonpe) [reply examples/bert/google_bert_qnn.py]: fixed now with script_dir flag

File: examples/clip/openai_clip-vit-base-patch16_ptq_qnn_qdq.json (hunk 6 above)
Comment (gtonpe) [reply examples/clip/openai_clip-vit-base-patch16_ptq_qnn_qdq.json]: File updated. Thank you for the suggestion

**Comment by justinchuby** [comment]
Could you rebase from main and make sure `lintrunner -a` passes, as we just updated the project to bump minimum python version to 3.9, which introduced a few style changes? Thanks

**Comment by jambayk** [comment]
Closed since #1775 has been merged
